File: Final/Code/Project_4/BaiToan5/Ex1_6/ex1_6.py
from itertools import permutations

def enumerate_perfect_matchings(p, q):
    if p != q:
        print("Không tồn tại ghép cặp hoàn hảo vì p ≠ q.")
        return

    U = [f"u{i+1}" for i in range(p)]
    V = [f"v{i+1}" for i in range(q)]

    print(f"Tất cả các perfect matching trong K_{{{p},{q}}}:\n")

    count = 0
    for perm in permutations(V):
        matching = [(U[i], perm[i]) for i in range(p)]
        print("Matching", count + 1, ":", matching)
        count += 1

    print(f"\nTổng số perfect matching: {count}")

def main():
    p = int(input("Nhập số đỉnh bên trái (p): "))
    q = int(input("Nhập số đỉnh bên phải (q): "))
    enumerate_perfect_matchings(p, q)

if __name__ == "__main__":
    main()

File: Final/Code/Project_4/BaiToan5/Ex1_6/ex1_6.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void enumerate_perfect_matchings(int p, int q) {
    if (p != q) {
        cout << "Không tồn tại ghép cặp hoàn hảo vì p ≠ q.\n";
        return;
    }

    vector<string> U, V;
    for (int i = 1; i <= p; ++i) {
        U.push_back("u" + to_string(i));
        V.push_back("v" + to_string(i));
    }

    sort(V.begin(), V.end()); 

    int count = 0;
    do {
        cout << "Matching " << ++count << ": ";
        for (int i = 0; i < p; ++i) {
            cout << "(" << U[i] << "," << V[i] << ") ";
        }
        cout << endl;
    } while (next_permutation(V.begin(), V.end()));

    cout << "\nTổng số perfect matching: " << count << endl;
}

int main() {
    int p, q;
    cout << "Nhập số đỉnh bên trái (p): ";
    cin >> p;
    cout << "Nhập số đỉnh bên phải (q): ";
    cin >> q;

    enumerate_perfect_matchings(p, q);
    return 0;
}